presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java

```diff
@@ -177,8 +178,15 @@ public ConnectorPageSource createPageSource(ConnectorTransactionHandle transacti
             TypeManager typeManager,
             LoadingCache<RowExpressionCacheKey, RowExpression> rowExpressionCache)
     {
-        Set<HiveColumnHandle> interimColumns = ImmutableSet.<HiveColumnHandle>builder()
-                .addAll(layout.getPredicateColumns().values())
+        RowExpression optimizedRemainingPredicate = rowExpressionCache.getUnchecked(new RowExpressionCacheKey(layout.getRemainingPredicate(), session));
+
+        ImmutableSet.Builder<HiveColumnHandle> interimColumnsBuilder = ImmutableSet.builder();
```

Comment:
@bhhari Filter expressions that get optimized into `null` should be eliminated at the query planning phase. Why is this not happening? CC: @highker

Reply:
@bhhari There is another scenario that needs to be handled. An original expression may use 3 columns, but optimized one only one column. In this case, we need to scan only one column, not three.